### PR TITLE
azurerm_api_management_api - deprecate `sku` in favour of the `sku_name` property

### PR DESCRIPTION
**Pull request by jeffreyCline**
Continuation of addressing issue #1500, Flatten the `SKU` block into a single attribute and unifying the handling of `SKU` across all resources within the Azure Terraform provider.

Example:

```
resource "azurerm_api_management_api" "test" {
  name      = "api-management-1"
  sku_name  = "Developer_1"
}
```

### CODE REVIEW

**Comment by r0bnet** [comment]
Thanks for this. Without this feature it's a nightmare to create something like t-shirt sizes if you need to define 3 values for each size (for instance for Postgres). Hope this will find it's way into 2.0.0

File: azurerm/data_source_api_management.go [hunk 1]
```diff
@@ -39,21 +39,9 @@ func dataSourceApiManagementService() *schema.Resource {
 				Computed: true,
 			},
 
-			"sku": {
-				Type:     schema.TypeList,
+			"sku_name": {
```
Comment (katbyte) [inline azurerm/data_source_api_management.go]: This is a breaking change. We should retain sku until 2.0.

File: azurerm/resource_arm_api_management.go [hunk 2]
```diff
@@ -77,15 +80,28 @@ func resourceArmApiManagementService() *schema.Resource {
 								string(apimanagement.SkuTypePremium),
 							}, false),
 						},
+
 						"capacity": {
 							Type:         schema.TypeInt,
 							Required:     true,
-							ValidateFunc: validation.IntAtLeast(0),
+							ValidateFunc: validation.IntAtLeast(1),
```
Comment (katbyte) [inline azurerm/resource_arm_api_management.go]: How come this value changed?

File: azurerm/resource_arm_api_management.go [hunk 3]
```diff
@@ -77,15 +80,28 @@ func resourceArmApiManagementService() *schema.Resource {
...
 						},
 					},
 				},
 			},
 
+			"sku_name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"sku"},
+				ValidateFunc: azure.SkuNameStringInSlice([]string{
```
Comment (katbyte) [inline azurerm/resource_arm_api_management.go]: This function seems rather specific to this service?

File: azurerm/resource_arm_api_management.go [hunk 4]
```diff
@@ -366,6 +382,14 @@ func resourceArmApiManagementServiceCreateUpdate(d *schema.ResourceData, meta in
 	client := meta.(*ArmClient).apiManagementServiceClient
 	ctx := meta.(*ArmClient).StopContext
 
+	log.Printf("[INFO] validating SKU for API Management Service.")
```
Comment (katbyte) [inline azurerm/resource_arm_api_management.go]: this seems like a mistake?

File: azurerm/resource_arm_api_management.go [hunk 5]
```diff
@@ -562,8 +585,21 @@ func resourceArmApiManagementServiceRead(d *schema.ResourceData, meta interface{
 		}
 	}
 
-	if err := d.Set("sku", flattenApiManagementServiceSku(resp.Sku)); err != nil {
-		return fmt.Errorf("Error setting `sku`: %+v", err)
+	if skuName == "" {
```
Comment (katbyte) [inline azurerm/resource_arm_api_management.go]: As mentioned in other PR, we can just set both and simplify this logic

File: azurerm/resource_arm_api_management.go [hunk 6]
```diff
@@ -850,21 +886,49 @@ func flattenAzureRmApiManagementMachineIdentity(identity *apimanagement.ServiceI
 }
 
 func expandAzureRmApiManagementSku(d *schema.ResourceData) *apimanagement.ServiceSkuProperties {
```
Comment (katbyte) [inline azurerm/resource_arm_api_management.go]: Can we split this into

`expandAzureRmApiManagementSku()` and `expandAzureRmApiManagementSkuName()` the the logic to pick in the calling function?

File: azurerm/resource_arm_api_management_api_test.go [hunk 7]
```diff
@@ -33,6 +33,43 @@ func TestAccAzureRMApiManagementApi_basic(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "soap_pass_through", "false"),
+					resource.TestCheckResourceAttr(resourceName, "is_current", "true"),
+					resource.TestCheckResourceAttr(resourceName, "is_online", "false"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_api_test.go]: We shouldn't be doing this.

File: azurerm/resource_arm_api_management_api_test.go [hunk 8]
```diff
@@ -59,6 +96,10 @@ func TestAccAzureRMApiManagementApi_wordRevision(t *testing.T) {
 				ResourceName:      resourceName,
 				ImportState:       true,
 				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_api_test.go]: We shouldn't be doing this.

File: azurerm/resource_arm_api_management_api_test.go [hunk 9]
```diff
@@ -113,6 +154,10 @@ func TestAccAzureRMApiManagementApi_soapPassthrough(t *testing.T) {
...
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_api_test.go]: We shouldn't be doing this.

File: azurerm/resource_arm_api_management_api_test.go [hunk 10]
```diff
@@ -139,6 +184,8 @@ func TestAccAzureRMApiManagementApi_importSwagger(t *testing.T) {
 				ImportState:       true,
 				ImportStateVerify: true,
 				ImportStateVerifyIgnore: []string{
+					"sku",
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_api_test.go]: We shouldn't be doing this.

File: azurerm/resource_arm_api_management_api_test.go [hunk 11]
```diff
@@ -168,6 +215,8 @@ func TestAccAzureRMApiManagementApi_importWsdl(t *testing.T) {
...
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_api_test.go]: And so on

File: azurerm/resource_arm_api_management_test.go [hunk 12]
```diff
@@ -30,6 +31,58 @@ func TestAccAzureRMApiManagement_basic(t *testing.T) {
 				ResourceName:      resourceName,
 				ImportState:       true,
 				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
```
Comment (katbyte) [inline azurerm/resource_arm_api_management_test.go]: We shouldn't be doing this.

File: website/docs/r/api_management.html.markdown [hunk 13]
```diff
@@ -46,7 +43,9 @@ The following arguments are supported:
 
 * `publisher_email` - (Required) The email of publisher/company.
 
-* `sku` - (Required) A `sku` block as documented below.
+* `sku`  - (Required) A `sku` block as documented below
```
Comment (katbyte) [inline website/docs/r/api_management.html.markdown]: Can we mark this as deprecated?

File: azurerm/resource_arm_api_management.go (hunk 2 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: Per the SDK implementation:
```
// Capacity - Capacity of the SKU (number of deployed units of the SKU). The default value is 1.
```

File: azurerm/resource_arm_api_management.go (hunk 3 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: There are lots of `SKU`'s that have the format of `Name_Capacity` so I will be able to reuse this function in other resources flatten `SKU` PR's. I have updated the name so it is more descriptive of what it is actually doing.

File: azurerm/resource_arm_api_management.go (hunk 4 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: Removed the line of code...

File: azurerm/resource_arm_api_management.go (hunk 5 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: I now set both and removed the logic...

File: azurerm/resource_arm_api_management.go (hunk 6 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: Done.

File: azurerm/data_source_api_management.go (hunk 1 above)
Comment (jeffreyCline) [reply azurerm/data_source_api_management.go]: kk

File: azurerm/resource_arm_api_management_api_test.go (hunk 9 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_api_test.go]: Fixed.

File: azurerm/resource_arm_api_management_api_test.go (hunk 10 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_api_test.go]: Fixed.

File: azurerm/resource_arm_api_management_test.go (hunk 12 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_test.go]: Fixed.

File: azurerm/resource_arm_api_management_api_test.go (hunk 7 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_api_test.go]: Fixed.

File: azurerm/resource_arm_api_management_api_test.go (hunk 8 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_api_test.go]: Fixed.

File: azurerm/resource_arm_api_management_api_test.go (hunk 11 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management_api_test.go]: Fixed.

File: website/docs/r/api_management.html.markdown (hunk 13 above)
Comment (jeffreyCline) [reply website/docs/r/api_management.html.markdown]: Fixed.

File: azurerm/data_source_api_management.go (hunk 1 above)
Comment (katbyte) [reply azurerm/data_source_api_management.go]: It appears your still removing the old property

File: azurerm/resource_arm_api_management.go (hunk 2 above)
Comment (katbyte) [reply azurerm/resource_arm_api_management.go]: that says default not minimum? thou it does make sense

**Comment by jeffreyCline** [comment]
Added `SKU` back into the DataSource...

File: website/docs/r/api_management.html.markdown [hunk 14]
```diff
@@ -46,7 +43,9 @@ The following arguments are supported:
 
 * `publisher_email` - (Required) The email of publisher/company.
 
-* `sku` - (Required) A `sku` block as documented below.
+* `sku`  - (Deprecated) A `sku` block as documented below
+
+* `sku_name` - (Required) `sku_name` is a string consisting of two parts separated by an underscore(_). The fist part is the `name`, valid values include: `Developer`, `Basic`, `Standard` and `Premium`. The second part is the `capacity` (e.g. the number of deployed units of the `sku`), which must be a positive `integer` (e.g. `Developer_1`).
```
Comment (katbyte) [inline website/docs/r/api_management.html.markdown]: You'll need to escape the _ as its all in italics now? or is it just the guthub viewer here

File: website/docs/r/api_management.html.markdown [hunk 15]
```diff
@@ -185,12 +184,14 @@ A `security` block supports the following:
...
 
-A `sku` block supports the following:
+A `sku` block supports the following: (Deprecated)
 
-* `name` - (Required) Specifies the Pricing Tier for the API Management Service. Possible values include: `Developer`, `Basic`, `Standard` and `Premium`.
+* `name` - (Required) Specifies the Pricing Tier for the API Management Service. Possible values include: Developer, Basic, Standard and Premium.
 
 * `capacity` - (Required) Specifies the Pricing Capacity for the API Management Service.
 
+-> **Note:** This property has been deprecated in favour of the `sku_name` property and will be removed in version 2.0 of the provider.
```
Comment (katbyte) [inline website/docs/r/api_management.html.markdown]: Could we add this to the upgrading to 2.0 guide found in the guides directory?

File: website/docs/r/api_management.html.markdown (hunk 14 above)
Comment (jeffreyCline) [reply website/docs/r/api_management.html.markdown]: Fixed.

File: website/docs/r/api_management.html.markdown (hunk 15 above)
Comment (jeffreyCline) [reply website/docs/r/api_management.html.markdown]: Sure... added both the data source and the resource.

File: azurerm/resource_arm_api_management.go (hunk 2 above)
Comment (jeffreyCline) [reply azurerm/resource_arm_api_management.go]: Reverted code and made it like it was, to avoid potential impact to existing users...

**Pull request merged**